### PR TITLE
Add gRPC interface to update/delete optional named vectors

### PR DESCRIPTION
**Pull request by timvisee**
Depends on <https://github.com/qdrant/qdrant/pull/1835>. Continuation of <https://github.com/qdrant/qdrant/pull/1758>. Part of <https://github.com/qdrant/qdrant/issues/1045>.

This PR adds a public and internal gRPC interface for updating and deleting optional named vectors.

Sorry for the PR size. There's a lot of layers to go through from the public interface to the internal functions. That makes it hard to shrink it further. Though big, the PR is quite simple. I'd be happy to walk through this in a call.

The basic interface consists of two actions:

https://github.com/qdrant/qdrant/blob/e0649929fc05410756fb2fdf3c851c608f4b1096/lib/api/src/grpc/proto/points.proto#L74-L88

I'd like to ask to take a look at the following with extra care:
- the public API design ([rendered](https://github.com/qdrant/qdrant/blob/e0649929fc05410756fb2fdf3c851c608f4b1096/docs/grpc/docs.md)) (discussed in a call before)
- the functions mutating segments to update/delete vectors

These features will be added in a separate PR to not grow this any bigger:
  - REST interface (<https://github.com/qdrant/qdrant/pull/1836>)
  - Integration tests

I did some basic testing for this against the gRPC interface locally.

### Tasks
- [x] Merge https://github.com/qdrant/qdrant/pull/1835
- [x] Rebase on `dev`
- [x] Undraft PR

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### CODE REVIEW

File: lib/segment/src/segment.rs [hunk 1]
```diff
@@ -113,6 +124,35 @@ impl Segment {
...
+    /// - new named vectors are inserted
+    /// - existing named vectors are replaced
+    /// - existing named vectors not specified are untouched and kept as-is
+    ///
+    /// This differs with [`replace_all_vectors`], because this keeps unspecified vectors as-is.
+    ///
+    /// # Warning
+    ///
+    /// Available for appendable segments only.
+    fn update_vectors(
```
Comment (timvisee) [inline lib/segment/src/segment.rs:140]: It is probably good to look into this function with extra care.

File: lib/segment/src/segment.rs [hunk 2]
```diff
@@ -766,6 +806,25 @@ impl SegmentEntry for Segment {
         }
     }
 
+    fn update_vectors(
```
Comment (timvisee) [inline lib/segment/src/segment.rs:809]: It is probably good to look into this function with extra care.

File: lib/segment/src/segment.rs [hunk 3]
```diff
@@ -766,6 +806,25 @@ impl SegmentEntry for Segment {
...
+        &mut self,
+        op_num: SeqNumberType,
+        point_id: PointIdType,
+        vectors: NamedVectors,
+    ) -> OperationResult<bool> {
+        let internal_id = self.id_tracker.borrow().internal_id(point_id);
+        match internal_id {
+            // Point does already not exist anymore
+            None => Ok(false),
```
Comment (generall) [inline lib/segment/src/segment.rs:818]: I think we should 404 here

Comment (timvisee) [reply lib/segment/src/segment.rs:818]: Good point!

This originates from deleting points, where it did make sense.

Comment (timvisee) [reply lib/segment/src/segment.rs]: https://github.com/qdrant/qdrant/blob/39d093902216afd015727e941e1c58c462912ea7/lib/segment/src/segment.rs#L817-L819

Done for both `update_vectors` and `delete_vectors`.

File: lib/segment/src/segment.rs (hunk 2 above)
Comment (generall) [reply lib/segment/src/segment.rs:809]: lgtm. I would expect it to be exactly as update of payload

File: lib/api/src/grpc/proto/points.proto [hunk 4]
```diff
@@ -71,6 +71,22 @@ message GetPoints {
   optional ReadConsistency read_consistency = 6; // Options for specifying read consistency guarantees
 }
 
+message UpdatePointVectors {
+  string collection_name = 1; // name of the collection
+  optional bool wait = 2; // Wait until the changes have been applied?
+  PointId id = 3; // ID to update vectors for
```
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/api/src/grpc/proto/points.proto]: Changed this to support a list of points.

**Review by generall — APPROVED** [review]

**Pull request merged**